rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,121 @@
+shepherd-leader: @kevincox 
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `<attrset>.[ <attrnames> ]`,
+which constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+attrs.[ a b c ]
+```
+
+This expression is syntactic sugar for:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+As the token `.` immediately preceding `[` is currently a syntax error,
+a Nix interpreter which supports this new language feature will be compatible
+with existing Nix code.
+
+This RFC is implemented here: https://github.com/nixos/nix/pull/5402
+
+For MVP functionality, only minor additions to `src/libexpr/parser.y` are
+needed. If accepted, the changes to the Nix interpreter can be backported
+to current releases if desired. Relevant language documentation and
+third-party parsers/linters would also need to be updated.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This would be useful for many languages and frameworks in Nixpkgs which
+extract packages from a package set argument.
+
+For example, `python3.withPackages (ps: ps.[ ... ])` will serve as a
+more fine-grained alternative to `python3.withPackages (ps: with ps; [ ... ])`.
+This would apply similarly to `vim.withPlugins`, `lua.withPackages`, etc.
+
+Certain list-typed `meta` fields could also make use of this feature, e.g.:
+```
+meta.licenses = lib.licenses.[ bsd3 mit ];
+meta.maintainers = lib.maintainers.[ johndoe janedoe ];
+```
+
```

Comment:
Is using it inside lists allowed?

```
[ [ 1 2 ] a.[ b c ] ]
```

:+1: or :-1:?

Reply:
Is using it inside of a inherit-as-list allowed?

```
a.[ b c d.[ e f ] ]
```

:+1: or :-1:?

Reply:
Is list inside of inherit-as-list allowed?

```
a.[ [ b c ] [ e f ] ]
```

:+1: or :-1:?

Reply:
Any more examples of practical cases that should be allowed?

Reply:
It isn't clear to me what that would mean so I'm voting no.

    a.[ [ b c ] [ e f ] ]

This seems the clearest and my best guess would be `[ [ a.b a.c ] [ a.e a.f ] ]`. But I still feel unsure about it. I would definitely not include this for the first version and we can always consider adding it later once we have more experience with the feature.

Reply:
I indeed agree on that one.

I've expanded the examples a bit with what they represent without 'inherit as list' syntax.

The other cases are a bit more tricky I think. You'd expect that you can add expressions to lists. A list in a list is possible (`flatten [ [ pkgs.python pkgs.jq ] ]`), so why wouldn't `flatten [ pkgs.[ python jq ] ]` work?

That said, if we do allow the syntax inside lists, because they are arbitrary expressions , just like lists are, then it would only make sense to allow it inside inherit-as-list.

Then again, I can't think of practical use-cases to allow these cases :shrug:

For the implementation, I can imagine that the parsing will be a bit more annoying, because it cannot be implemented as an arbitrary expression. It needs explicit checks whether it is used inside lists or inherit-as-lists.

Reply:
I would support inherit-as-list as expression, while its content must be expressions but with restrictions (well, restrictions are needed in any case there…)